### PR TITLE
continue to add type system for batch write

### PR DESCRIPTION
**Pull request by marsishandsome**
1. add to TIMESTAMP
    2. add to DATE
    3. add to DATETIME
    4. add to STRING
    5. add to DECIMAL
    6. add to BIT
    7. add to BYTES
    8. add to ENUM

### CODE REVIEW

**Comment by marsishandsome** [comment]
@zhexuany @birdstorm PTAL

File: core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala [hunk 1]
```diff
@@ -277,6 +377,162 @@ object TiConverter {
...
+      case v: java.lang.String => parseEnumName(targetColumnInfo.getType.getElems, v)
+      case _ =>
+        val r: java.lang.Long = convertToUnsigned(targetColumnInfo, value)
+        parseEnumValue(targetColumnInfo.getType.getElems, r.toInt)
+    }
+
+    result
+  }
+
+  private def parseEnumName(elems: java.util.List[java.lang.String],
```
Comment (zhexuany) [inline core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala]: better move to java client side.

File: core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala [hunk 2]
```diff
@@ -277,6 +377,162 @@ object TiConverter {
...
+      java.lang.Integer.parseInt(name)
+    } catch {
+      case _: Throwable =>
+        throw new TiBatchWriteException(s"item $name is not in enum $elems")
+    }
+
+    parseEnumValue(elems, num)
+  }
+
+  private def parseEnumValue(elems: java.util.List[java.lang.String],
```
Comment (zhexuany) [inline core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala]: ditto.

**Comment by marsishandsome** [comment]
@birdstorm @zhexuany PTAL

File: core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala (hunk 1 above)
Comment (marsishandsome) [reply core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala]: fixed

File: core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala (hunk 2 above)
Comment (marsishandsome) [reply core/src/main/scala/com/pingcap/tispark/utils/TiConverter.scala]: fixed

File: core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala [hunk 3]
```diff
@@ -498,14 +498,14 @@ class TiBatchWrite(@transient val df: DataFrame,
...
-          var data = sparkRow.get(i)
+          var data = TiConverter.sparkSQLObjectToJavaObject(sparkRow.get(i))
           val sparkDataType = sparkRow.schema(i).dataType
           val colName = sparkRow.schema(i).name
           // check do we need fill auto increment column
           if (colName.equals(autoincrementCol.getName)) {
             // only update auto increment column when the value is null
             if (data == null) {
-              data = handleId
+              data = new java.lang.Long(handleId)
```
Comment (birdstorm) [inline core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala]: Long.box()

File: tikv-client/src/main/java/com/pingcap/tikv/types/BitType.java [hunk 4]
```diff
@@ -36,6 +38,18 @@ protected BitType(TiColumnInfo.InternalTypeHolder holder) {
     super(holder);
   }
 
+  @Override
+  protected Object doConvertToTiDBType(Object value)
+      throws ConvertNotSupportException, ConvertOverflowException {
+    Long result = Converter.safeConvertToUnsigned(value, this.unsignedUpperBound());
+    long targetLength = this.getLength();
+    long upperBound = 1 << targetLength;
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/types/BitType.java]: should this be `1L << targetLength`?

Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/BitType.java]: fix

File: core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala (hunk 3 above)
Comment (marsishandsome) [reply core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala]: fixed

File: tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java [hunk 5]
```diff
@@ -47,6 +49,28 @@ public Date getOriginDefaultValueNonNull(String value, long version) {
...
+  private java.sql.Date convertToMysqlDate(Object value) throws ConvertNotSupportException {
+    java.sql.Date result;
+    if (value instanceof Long) {
+      result = new java.sql.Date((Long) value);
+    } else if (value instanceof String) {
+      result = java.sql.Date.valueOf((String) value);
+    } else if (value instanceof java.sql.Date) {
+      result = (java.sql.Date) value;
+    } else if (value instanceof java.sql.Timestamp) {
+      result = new java.sql.Date(((java.sql.Timestamp) value).getTime());
```
Comment (zhexuany) [inline tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: We may consider the effect of timezone.

File: tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java [hunk 6]
```diff
@@ -61,6 +63,25 @@ private static BigDecimal unsignedValueOf(long x) {
...
 
+  @Override
+  protected Object doConvertToTiDBType(Object value)
+      throws ConvertNotSupportException, ConvertOverflowException {
+    // TODO: support write to YEAR
+    if (this.getType() == MySQLType.TypeYear) {
+      throw new ConvertNotSupportException(value.getClass().getName(), this.getClass().getName());
+    }
+
+    Long result;
```
Comment (zhexuany) [inline tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java]: what if we want to convert a non-number object to long? e.g "1000", How do we handle this?

File: tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java [hunk 7]
```diff
@@ -25,7 +24,7 @@ private static TiTableInfo createTable() {
         .addColumn("c1", IntegerType.INT, true)
         .addColumn("c2", IntegerType.BIGINT)
         .addColumn("c3", DateTimeType.DATETIME)
-        .addColumn("c4", TimestampType.TIMESTAMP)
+        // .addColumn("c4", TimestampType.TIMESTAMP)
```
Comment (zhexuany) [inline tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java]: Should we remove this commented code?

File: tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java [hunk 8]
```diff
@@ -41,7 +40,7 @@ private void makeValues() {
     values.add(1L);
     DateTime dateTime = DateTime.parse("1995-10-10");
     values.add(new Timestamp(dateTime.getMillis()));
-    values.add(new Timestamp(dateTime.getMillis()));
+    // values.add(new Timestamp(dateTime.getMillis()));
     values.add("abc");
     values.add("中");
```
Comment (zhexuany) [inline tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java]: ditto.

File: tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java (hunk 6 above)
Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java]: see UT `Test Convert from String to SIGNED`

File: tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java (hunk 7 above)
Comment (marsishandsome) [reply tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java]: currently only support write `java.lang.Long` to `TIMESTAMP`

File: tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java (hunk 5 above)
Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: would you please provide a test case, which is wrong for the current code?

File: tikv-client/src/main/java/com/pingcap/tikv/types/BitType.java [hunk 9]
```diff
@@ -36,6 +38,18 @@ protected BitType(TiColumnInfo.InternalTypeHolder holder) {
     super(holder);
   }
 
+  @Override
+  protected Object doConvertToTiDBType(Object value)
+      throws ConvertNotSupportException, ConvertOverflowException {
+    Long result = Converter.safeConvertToUnsigned(value, this.unsignedUpperBound());
+    long targetLength = this.getLength();
+    long upperBound = 1L << targetLength;
+    if (targetLength < 64L && java.lang.Long.compareUnsigned(result, upperBound) >= 0) {
```
[the remark this answers is not in the document]
Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/BitType.java]: agree

File: core/src/main/resources/log4j.properties.template [hunk 10]
```diff
@@ -1,26 +1,26 @@
 # Set everything to be logged to the console
-log4j.rootCategory=INFO, console
+log4j.rootCategory=ERROR, console
```
[the remark this answers is not in the document]
Comment (marsishandsome) [reply core/src/main/resources/log4j.properties.template]: agree

File: tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java (hunk 5 above)
Comment (birdstorm) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: do we have unit tests for `convertToMysqlDate()`?

Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: disable all timezone related conversion

File: tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java (hunk 7 above)
Comment (zhexuany) [reply tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java]: In that case, a TODO can add here or elsewhere associated with this problem.

File: tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java (hunk 5 above)
Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: @birdstorm  see `ToDateSuite`

File: tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java (hunk 7 above)
Comment (marsishandsome) [reply tikv-client/src/test/java/com/pingcap/tikv/codec/TableCodecTest.java]: agree

File: tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java (hunk 5 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: Gotcha. Let me fix timezone related issue = =

Comment (marsishandsome) [reply tikv-client/src/main/java/com/pingcap/tikv/types/DateType.java]: awesome

**Pull request merged**